src/server/cluster/cluster_family.cc

```diff
@@ -178,25 +178,53 @@ void ClusterFamily::ClusterInfo(ConnectionContext* cntx) {
   auto append = [&msg](absl::AlphaNum a1, absl::AlphaNum a2) {
     absl::StrAppend(&msg, a1, ":", a2, "\r\n");
   };
-  // info command just return some stats about this instance
-  int known_nodes = 1;
+
+  // Initialize response variables to emulated mode.
+  string_view state = "ok"sv;
+  SlotId slots_assigned = ClusterConfig::kMaxSlotNum + 1;
+  size_t known_nodes = 1;
   long epoch = 1;
-  ServerState& etl = *ServerState::tlocal();
-  if (etl.is_master) {
-    auto vec = server_family_->GetDflyCmd()->GetReplicasRoleInfo();
-    if (!vec.empty()) {
-      known_nodes = 2;
+  size_t cluster_size = 1;
+
+  if (is_emulated_cluster_) {
+    ServerState& etl = *ServerState::tlocal();
+    if (etl.is_master) {
+      auto vec = server_family_->GetDflyCmd()->GetReplicasRoleInfo();
+      if (!vec.empty()) {
+        known_nodes = 2;
+      }
+    } else {
+      if (server_family_->HasReplica()) {
+        known_nodes = 2;
+        epoch = server_family_->GetReplicaInfo().master_last_io_sec;
+      }
     }
+  } else if (!cluster_config_->IsConfigured()) {
+    state = "fail"sv;
+    slots_assigned = 0;
+    cluster_size = 0;
+    known_nodes = 0;
   } else {
-    if (server_family_->HasReplica()) {
-      known_nodes = 2;
-      epoch = server_family_->GetReplicaInfo().master_last_io_sec;
+    absl::flat_hash_set<string_view> nodes;
+    absl::flat_hash_set<string_view> master_nodes;
+    auto config = cluster_config_->GetConfig();
+    for (const auto& shard_config : config) {
+      nodes.insert(shard_config.master.id);
+      if (!shard_config.slot_ranges.empty()) {
+        master_nodes.insert(shard_config.master.id);
+      }
+
+      for (const auto& replica : shard_config.replicas) {
+        nodes.insert(replica.id);
+      }
     }
+    known_nodes = nodes.size();
+    cluster_size = master_nodes.size();
   }
-  int cluster_size = known_nodes - 1;
-  append("cluster_state", "ok");
-  append("cluster_slots_assigned", ClusterConfig::kMaxSlotNum);
-  append("cluster_slots_ok", ClusterConfig::kMaxSlotNum);
+
+  append("cluster_state", state);
+  append("cluster_slots_assigned", slots_assigned);
```

Comment:
I see that before in emulated mode slots_assigned is ClusterConfig::kMaxSlotNum
now it is ClusterConfig::kMaxSlotNum + 1
What should be the value?

Reply:
Yes, this is a bug fix that I introduced in the last PR :|

From: https://redis.io/commands/cluster-info/

> cluster_slots_assigned: [...] This number should be 16384 for the node to work properly [...]

It's my fault, I'm used to the notation where max is 1-post the end, not the last one.

On a similar subject, I also changed one more thing in the response: in emulated mode, cluster_size used to return 0 if there were no replicas. I think that it was a(n old) bug.